userland/libtock/ble.c

```diff
@@ -1,43 +1,123 @@
+/*
+ * BLE setup functions
+ */
+
 #include "ble.h"
+#include "tock.h"
+#include <stdbool.h>
+#include <stdint.h>
 #include <stdio.h>
 #include <string.h>
 
-int ble_adv_set_txpower(BLE_TX_Power_t power) {
-  return command(DRIVER_RADIO, BLE_CFG_TX_POWER, power);
+#define GAP_DATA_MAX_SIZE 31
+
+
+
+// helper functions
+static int s_ble_cfg_addr (uint8_t *addr, uint8_t addr_len);
+static int s_ble_cfg_flags (uint8_t flags);
+static int s_ble_cfg_adv_itv (uint16_t advertising_itv_ms);
+static int s_ble_cfg_gap_data (GapAdvertisementData_t header, uint8_t *data, uint8_t len);
+
+// local VARIABLES
+static uint8_t s_gap[GAP_DATA_MAX_SIZE];
+
+
+/*******************************************************************************
+ *   USER-SPACE BLE API
+ ******************************************************************************/
+
+int ble_initialize(uint16_t advertising_itv_ms, uint8_t flags, uint8_t *addr,
+                   uint8_t addr_len) {
+  int err;
+
+  // configure advertisement interval
+  // if the interval is less than 20 or bigger than 10240 to kernel
+  // will use default value
+  err = s_ble_cfg_adv_itv(advertising_itv_ms);
+  if (err < TOCK_SUCCESS) return err;
+
+  // configure advertisement flags in the packet
+  if (flags != 0x00) {
+    err = s_ble_cfg_flags(flags);
+    if (err < TOCK_SUCCESS) return err;
+  }
+
+  // if this is NULL advertisment address {0, 0, 0, 0, 0, 0}  will be used
+  if (addr != NULL) {
+    err = s_ble_cfg_addr(addr, addr_len);
+    if (err < TOCK_SUCCESS) return err;
+  }
+
+  return err;
 }
 
-int ble_adv_set_interval(uint16_t interval) {
-  return command(DRIVER_RADIO, BLE_CFG_ADV_INTERVAL, interval);
+int ble_start_advertising(void) {
+  return command(BLE_DRIVER_NUMBER, BLE_ADV_START_CMD, 0);
 }
 
-int ble_adv_data(uint8_t type, const unsigned char *data, uint8_t len) {
-  return allow(DRIVER_RADIO, type, (void*)data, len);
+int ble_stop_advertising(void) {
+  return command(BLE_DRIVER_NUMBER, BLE_ADV_STOP_CMD, 1);
 }
 
-int ble_adv_clear_data(void){
-  return command(DRIVER_RADIO, BLE_ADV_CLEAR_DATA, 1);
+int ble_reset_advertisement(void) {
+  return command(BLE_DRIVER_NUMBER, BLE_ADV_CLEAR_DATA_CMD, 1);
 }
 
-int ble_adv_start(BLE_Gap_Mode_t mode) {
-  return command(DRIVER_RADIO, BLE_ADV_START, mode);
+int ble_advertise_name(uint8_t *device_name, uint8_t len) {
+  return s_ble_cfg_gap_data(GAP_COMPLETE_LOCAL_NAME, device_name, len);
 }
 
-int ble_adv_scan(const unsigned char *data, uint8_t len, subscribe_cb callback) {
+int ble_advertise_uuid16(uint16_t *uuid16, uint8_t len) {
+  return s_ble_cfg_gap_data(GAP_COMPLETE_LIST_16BIT_SERVICE_IDS, (uint8_t*)uuid16, len);
+}
+
+int ble_advertise_service_data(uint16_t uuid16, uint8_t *data, uint8_t data_len) {
+  // buffer overflow in libtock don't allow
+  if (data_len + 2 > GAP_DATA_MAX_SIZE) {
+    return TOCK_FAIL;
+  }else {
+    memset(s_gap, 0, GAP_DATA_MAX_SIZE);
```

Comment:
Why not make ` s_gap` as a local buffer on the stack in this function instead of a file variable?

Reply:
Sure.